packages/core/app/src/initialize.ts

```diff
@@ -1,73 +1,105 @@
-        // unregistered by reusing a reference to the wallet to obtain the unregister function for it.
-        wallets = wallets.filter((wallet) => !registered.includes(wallet));
-        // If there are no new wallets to register, just return a no-op unregister function.
-        // eslint-disable-next-line @typescript-eslint/no-empty-function
-        if (!wallets.length) return () => {};
-
-        registered.push(...wallets);
-        listeners['register']?.forEach((listener) => listener(...wallets));
-        // Return a function that unregisters the registered wallets.
-        return function unregister(): void {
-            wallets
-                .map((wallet) => registered.indexOf(wallet))
-                .filter((index) => index !== -1)
-                .sort()
-                .reverse()
-                .forEach((index) => registered.splice(index, 1));
-            listeners['unregister']?.forEach((listener) => listener(...wallets));
-        };
-    }
+/** TODO: docs */
+export type WalletsEventNames = keyof WalletsEvents;
 
-    function get(): ReadonlyArray<Wallet> {
-        // Return a copy so the registered wallets can't be referenced or mutated.
-        return registered.slice();
-    }
+const registered = new Set<Wallet>();
+const listeners: { [E in WalletsEventNames]?: WalletsEvents[E][] } = {};
 
-    function on<E extends WalletsEventNames>(event: E, listener: WalletsEvents[E]): () => void {
-        listeners[event]?.push(listener) || (listeners[event] = [listener]);
-        // Return a function that removes the event listener.
-        return function off(): void {
-            listeners[event] = listeners[event]?.filter((existingListener) => listener !== existingListener);
-        };
+function push(...callbacks: ReadonlyArray<WalletsCallback>): void {
+    for (const callback of callbacks) {
+        callback({ register });
     }
+}
+
+function register(...wallets: ReadonlyArray<Wallet>): () => void {
+    // Filter out wallets that have already been registered.
+    // This prevents the same wallet from being registered twice, but it also prevents wallets from being
+    // unregistered by reusing a reference to the wallet to obtain the unregister function for it.
+    wallets = wallets.filter((wallet) => !registered.has(wallet));
+    // If there are no new wallets to register, just return a no-op unregister function.
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    if (!wallets.length) return () => {};
+
+    wallets.forEach((wallet) => registered.add(wallet));
+    listeners['register']?.forEach((listener) => listener(...wallets));
+    // Return a function that unregisters the registered wallets.
+    return function unregister(): void {
+        wallets.forEach((wallet) => registered.delete(wallet));
+        listeners['unregister']?.forEach((listener) => listener(...wallets));
+    };
```

Comment:
The `registered` array has been replaced with a Set, so we no longer need the fun but complicated unregister logic.